cardinal/chain/adapter.go

```diff
@@ -17,8 +18,18 @@ import (
 
 // Adapter is a type that helps facilitate communication with the EVM base shard.
 type Adapter interface {
-	Submit(ctx context.Context, namespace string, tick uint64, txs []byte) error
-	QueryBatches(ctx context.Context, req *shardtypes.QueryBatchesRequest) (*shardtypes.QueryBatchesResponse, error)
+	Writer
+	Reader
+}
+
+type Writer interface {
+	Submit(ctx context.Context, p *sign.SignedPayload, txID, tick uint64) error
+}
+
+type Reader interface {
+	QueryTransactions(
+		context.Context,
+		*shardtypes.QueryTransactionsRequest) (*shardtypes.QueryTransactionsResponse, error)
```

Comment:
I trust that this is correct, but wondering: why have we chosen to use pointers in these vars?

Reply:
gRPC client/server generated code operates on pointer requests/response types. since this is mostly an alias to the underlying gRPC call, i just kept the params the same. they could be value types instead of pointers, but don't think it matters too much